### PR TITLE
Clarify rules for copies vs views when indexing

### PR DESCRIPTION
**Pull request by shoyer**
This mostly reorganizes/updates the indexing documentation.

It also catches an edge case where a slice based indexer could be converted silently into array indexer. This is a strange edge case involving unsorted or duplicate indexes that arguably pandas shouldn't allow anyways.

### CODE REVIEW

**Comment by shoyer** [comment]
@jhamman any thoughts?

File: doc/indexing.rst [hunk 1]
```diff
@@ -191,39 +194,126 @@ index labels along a dimension dropped:
...
+.. _nearest neighbor lookups:
 
-Indexing details
-----------------
+Nearest neighbor lookups
+------------------------
 
-Like pandas, whether array indexing returns a view or a copy of the underlying
-data depends entirely on numpy:
+The label based selection methods :py:meth:`~xray.Dataset.sel`,
```
Comment (jhamman) [inline doc/indexing.rst]: also `sel_points`

File: doc/indexing.rst [hunk 2]
```diff
@@ -191,39 +194,126 @@ index labels along a dimension dropped:
...
+* Positional indexing with only integers and slices returns a view.
+* Positional indexing with arrays or lists returns a copy.
+
+The rules for label based indexing are more complex:
+
+* Label-based indexing with only slices returns a view.
+* Label-based indexing with arrays returns a copy.
+* Label-based indexing with scalars returns a view or a copy, depending
+  upon if the corresponding positional indexer can be represented as an
+  integer or a slice object.
```
Comment (jhamman) [inline doc/indexing.rst]: for the last point, an example showing both behaviors may be helpful.

File: doc/indexing.rst [hunk 3]
```diff
@@ -242,6 +332,10 @@ original values are subset to the index labels still found in the new labels,
 and values corresponding to new labels not found in the original object are
 in-filled with `NaN`.
 
+Xray operations that combine multiple xray objects generally automatically
+align their arguments. However, manual alignment can be useful for greater
```
Comment (jhamman) [inline doc/indexing.rst]: do you mean indexes not arguments.  This line doesn't make much sense to me.

**Comment by jhamman** [comment]
I made a few comments. I think you'll have to sort out the merge conflicts coming from #507.

**Comment by jhamman** [comment]
I made a few comments. I think you'll have to sort out the merge conflicts coming from #507.

**Comment by shoyer** [comment]
Revised and rebased.

File: doc/indexing.rst (hunk 2 above)
Comment (shoyer) [inline doc/indexing.rst]: unfortunately, the distinction here is pretty complex and I'm not even sure I understand it fully...

File: doc/indexing.rst (hunk 3 above)
Comment (shoyer) [inline doc/indexing.rst]: revised, hopefully this is better

**Comment by jhamman** [comment]
I'm not able to build the docs from where I am but from what I can tell, this looks good.

**Pull request merged**